src/dshould/prettyprint.d

```diff
@@ -0,0 +1,385 @@
+
+    prettyprint("Foo()").should.equal("Foo()");
+    prettyprint("Foo(Bar(Baz()), Baq())", 16).should.equal(
+"Foo(
+    Bar(Baz()),
+    Baq()
+)");
+    prettyprint("Foo(Bar(Baz()), Baq())", 13).should.equal(
+"Foo(
+    Bar(
+        Baz()
+    ),
+    Baq()
+)");
+}
+
+private enum indent = "    ";
+
+private string prettyprint(const Tree tree, size_t width)
+{
+    import std.string : stripLeft;
+
+    Appender!string result;
+
+    // skip prefix so caller can decide whether or not to strip
+    void walkOneLine(const Tree tree)
+    {
+        if (tree.parenType.isNull)
+        {
+            return;
+        }
+        result ~= tree.parenType.get.opening;
+        tree.children.enumerate.each!((index, child) {
+            if (index > 0)
+            {
+                result ~= ",";
+            }
+            result ~= child.prefix;
+            walkOneLine(child);
+        });
+        result ~= tree.parenType.get.closing;
+    }
+
+    void walk(size_t level, const Tree tree)
+    {
+        if (!tree.lengthExceeds(width - level * indent.length))
+        {
+            result ~= tree.prefix.stripLeft;
+            walkOneLine(tree);
+            return;
+        }
+
+        result ~= tree.prefix.stripLeft;
+        if (tree.parenType.isNull)
+        {
+            return;
+        }
+        result ~= tree.parenType.get.opening;
+        tree.children.enumerate.each!((index, child) {
+            if (index > 0)
+            {
+                result ~= ",";
+            }
+            result ~= "\n";
+            (level + 1).iota.each!((_) { result ~= indent; });
+            walk(level + 1, child);
+        });
+        result ~= "\n";
+        level.iota.each!((_) { result ~= indent; });
+        result ~= tree.parenType.get.closing;
+    }
+    walk(0, tree);
+    return result.data;
+}
+
+private Nullable!Tree parse(string text)
+{
+    auto textRange = text.quoted;
+    auto result = parse(textRange);
+
+    if (!textRange.empty)
+    {
+        return Nullable!Tree();
+    }
+
+    return result;
+}
+
+@("parse a paren expression to a tree")
+unittest
+{
+    import dshould : equal, should;
+
+    parse("Foo").should.equal(Tree("Foo").nullable);
+    parse(`"Foo"`).should.equal(Tree(`"Foo"`).nullable);
+    parse("Foo,").should.equal(Nullable!Tree());
+    parse("Foo, Bar").should.equal(Nullable!Tree());
+    parse("Foo()").should.equal(Tree("Foo", ParenType.paren.nullable));
+    parse("Foo[a, b]").should.equal(Tree(
+        "Foo",
+        ParenType.squareBracket.nullable,
+        [Tree("a"), Tree(" b")]).nullable);
+    parse(`Foo{"\""}`).should.equal(Tree(
+        "Foo",
+        ParenType.curlyBracket.nullable,
+        [Tree(`"\""`)]).nullable);
+}
+
+private Nullable!Tree parse(ref QuotedText textRange)
+{
+    auto parenStart = textRange.findAmong("({[");
+    auto closer = textRange.findAmong(",)}]");
+
+    if (textRange.textUntil(closer).length < textRange.textUntil(parenStart).length)
```

Comment:
Is this "error handling"? Because there is no kind of error handling below: just return `Nullable!Tree()`.

Reply:
It's not.